src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/InputManager.cs

```diff
@@ -974,14 +951,14 @@ private void RequestContinueProcessingStagingArea()
         private ProcessInputEventArgs _processInputEventArgs;
         private PreProcessInputEventArgs _preProcessInputEventArgs;
 
-        //these four events introduced for secutiy purposes
-        private event PreProcessInputEventHandler _preProcessInput;
-
-        private event NotifyInputEventHandler _preNotifyInput;
-
-        private event NotifyInputEventHandler _postNotifyInput;
-
-        private event ProcessInputEventHandler _postProcessInput;
+        // These four events introduced for security purposes. Rather than just store the multicast delegate
+        // and have to do GetInvocationList on each invocation (in order to invoke delegates in reverse-registered
+        // order), we get the invocation list when the events are updated, and then use that cached list
+        // on every invocation.
+        private Tuple<PreProcessInputEventHandler, Delegate[]> _preProcessInput;
```

Comment:
How about ValueTuple?

Reply:
That can't be swapped atomically with an Interlocked.  I don't actually know if thread-safety is necessary here, but `event` provides that automatically, so I brought the behavior forward.

Reply:
@stephentoub  Yeah, thread safety is necessary. Such as I will add handle in my `Stylus Input` thread.